File: src/java/PrefixSumArray.java
// DSA09 (Java): Prefix Sum Array

import java.util.Scanner;

public class PrefixSumArray {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the number of elements: ");
        int n = sc.nextInt();

        int[] arr = new int[n];
        System.out.print("Enter the elements: ");
        for (int i = 0; i < n; i++)
            arr[i] = sc.nextInt();

        System.out.print("The array elements are: ");
        for (int i : arr)
            System.out.print(i + " ");
        System.out.println();

        int[] prefix = new int[n];
        prefixArray(arr, prefix);

        System.out.print("The prefix array: ");
        for (int i : prefix)
            System.out.print(i + " ");
        System.out.println();

        if (n >= 3)
            System.out.println("Range sum [0, 2]: " + rangeSum(prefix, 0, 2));

        if (n >= 4)
            System.out.println("Range sum [1, 3]: " + rangeSum(prefix, 1, 3));

        if (n >= 5)
            System.out.println("Range sum [2, 4]: " + rangeSum(prefix, 2, 4));

        sc.close();
    }

    static void prefixArray(int[] arr, int[] prefix) {
        int n = arr.length;

        prefix[0] = arr[0];

        for (int i = 1; i < n; i++)
            prefix[i] = prefix[i - 1] + arr[i];
    }

    static int rangeSum(int[] prefix, int left, int right) {
        if (left == 0)
            return prefix[right];

        return prefix[right] - prefix[left - 1];
    }
}

File: src/kotlin/PrefixSumArray.kt
// DSA09 (Kotlin): Prefix Sum Array

import java.util.Scanner

fun main() {
    val sc = Scanner(System.`in`)

    print("Enter the number of elements: ")
    val n = sc.nextInt()

    val arr = IntArray(n)
    print("Enter the elements: ")
    for (i in arr.indices)
        arr[i] = sc.nextInt()

    print("The array elements are: ")
    for (i in arr)
        print("$i ")
    println()

    val prefix = IntArray(n)
    prefixArray(arr, prefix)

    print("The prefix array: ")
    for (i in prefix)
        print("$i ")
    println()

    if (n >= 3)
        println("Range sum [0, 2]: ${rangeSum(prefix, 0, 2)}")

    if (n >= 4)
        println("Range sum [1, 3]: ${rangeSum(prefix, 1, 3)}")

    if (n >= 5)
       println("Range sum [2, 4]: ${rangeSum(prefix, 2, 4)}")

    sc.close()
}

fun prefixArray(arr: IntArray, prefix: IntArray) {
    val n = arr.size

    prefix[0] = arr[0]

    for (i in 1 until n)
        prefix[i] = prefix[i - 1] + arr[i]
}

fun rangeSum(prefix: IntArray, left: Int, right: Int): Int {
    if (left == 0)
        return prefix[right]

    return prefix[right] - prefix[left - 1]
}
